compiler/src/dotty/tools/dotc/typer/Checking.scala

```diff
@@ -625,6 +624,24 @@ trait Checking {
       case _ =>
     }
   }
+
+  /** Check that method parameter types do not reference their own parameter
+   *  or later parameters in the same parameter section.
+   */
+  def checkNoForwardDependencies(vparams: List[ValDef])(implicit ctx: Context): Unit = vparams match {
```

Comment:
One is a source check, the other an internal assertion that results in an exception. Not all method types are created directly from source, so the assertion is useful.